### PR TITLE
ElementR: Add `checkOwnCrossSigningTrust` to `CryptoApi`

### PR DESCRIPTION
**Pull request by florianduros**
## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md))

Content of the PR

- Add `checkOwnCrossSigningTrust` impl to rust-crypto
- Deprecated `MatrixClient.checkOwnCrossSigningTrust`
- The (unreadable) changes in `crypto.spec.ts` are documentation and function sharing. At one exception: the test `checkOwnCrossSigningTrust` is added
- Move `bootstrapSecretStorage` and `checkOwnCrossSigningTrust` into the same test suite to share utility functions

---
Here's what your changelog entry will look like:

## ✨ Features
 * ElementR: Add `checkOwnCrossSigningTrust` to `CryptoApi` ([\#3546](https://github.com/matrix-org/matrix-js-sdk/pull/3546)). Contributed by @florianduros.

### CODE REVIEW

**Review by germain-gg — APPROVED** [review]
Looks great, thank you!

File: src/rust-crypto/rust-crypto.ts [hunk 1]
```diff
@@ -640,6 +642,43 @@ export class RustCrypto extends TypedEventEmitter<RustCryptoEvents, RustCryptoEv
...
+        );
+
+        // Verify the device and upload the cross signing signatures
+        const request: RustSdkCryptoJs.SignatureUploadRequest = await device.verify();
+        await this.http.authedRequest(
+            Method.Post,
+            "/_matrix/client/v3/keys/signatures/upload",
+            undefined,
+            request.body,
+        );
```
Comment (richvdh) [inline src/rust-crypto/rust-crypto.ts:679]: surely this should be documented in the doc-comment in cryptoApi?

File: src/crypto-api.ts [hunk 2]
```diff
@@ -490,5 +501,15 @@ export interface GeneratedSecretStorageKey {
     encodedPrivateKey?: string;
 }
 
+/**
+ * Options object for {@link CryptoApi#checkOwnCrossSigningTrust}.
+ */
+export interface CheckOwnCrossSigningTrustOpts {
+    /**
+     * If true, will fetch the private cross signing keys from the secret storage
+     */
```
Comment (richvdh) [inline src/crypto-api.ts:510]: what happens if it is false?

File: src/rust-crypto/rust-crypto.ts [hunk 3]
```diff
@@ -640,6 +642,43 @@ export class RustCrypto extends TypedEventEmitter<RustCryptoEvents, RustCryptoEv
...
+     * Implementation of {@link CryptoApi#checkOwnCrossSigningTrust}
+     */
+    public async checkOwnCrossSigningTrust({ allowPrivateKeyRequests }: CheckOwnCrossSigningTrustOpts): Promise<void> {
+        // Download the device keys and add them into the olm machine
+        const queryResult = await this.downloadDeviceList(new Set([this.userId]));
+        await this.olmMachine.markRequestAsSent(
+            uuidv4(),
+            RustSdkCryptoJs.RequestType.KeysQuery,
+            JSON.stringify(queryResult),
+        );
```
Comment (richvdh) [inline src/rust-crypto/rust-crypto.ts:655]: why is this necessary? making up a request id and telling the rust layer that we have sent that made-up request ID seems like an odd thing to do.

File: src/crypto-api.ts [hunk 4]
```diff
@@ -229,6 +229,17 @@ export interface CryptoApi {
      */
     createRecoveryKeyFromPassphrase(password?: string): Promise<GeneratedSecretStorageKey>;
 
+    /**
+     * Check the cross signing trust of the current user
```
Comment (richvdh) [inline src/crypto-api.ts:233]: > Check the cross signing trust of the current user

what happens if the check fails?

**Review by richvdh — CHANGES_REQUESTED** [review]
I has questions

File: src/crypto-api.ts (hunk 2 above)
Comment (florianduros) [reply src/crypto-api.ts:510]: The cross signing keys are not imported

File: src/rust-crypto/rust-crypto.ts (hunk 1 above)
Comment (florianduros) [reply src/rust-crypto/rust-crypto.ts:679]: In the doc-comment: `Verify the current device`. We should also tell that we are uploading the device signatures ?

File: src/crypto-api.ts (hunk 4 above)
Comment (florianduros) [reply src/crypto-api.ts:233]: Depends when the failure occurs; the device will not be verified (the device signatures will not be uploaded), the cross signing keys or the device keys will no be imported

File: src/rust-crypto/rust-crypto.ts (hunk 1 above)
Comment (richvdh) [reply src/rust-crypto/rust-crypto.ts:679]: I would have thought so. "Verify" is quite different to "sign and upload".

File: src/crypto-api.ts (hunk 2 above)
Comment (richvdh) [reply src/crypto-api.ts:510]: ok, please could you make this explicit in the comment.

File: src/rust-crypto/rust-crypto.ts (hunk 3 above)
Comment (richvdh) [reply src/rust-crypto/rust-crypto.ts:655]: well, first of all, doesn't the OlmMachine download the device keys automatically?

File: src/crypto-api.ts (hunk 4 above)
Comment (richvdh) [reply src/crypto-api.ts:233]: please put this in the documentation

Comment (richvdh) [reply src/crypto-api.ts:233]: (will an exception be thrown if the check fails?)

File: src/rust-crypto/rust-crypto.ts (hunk 3 above)
Comment (florianduros) [reply src/rust-crypto/rust-crypto.ts:655]: Oh, you mean that I don't need the download the device list manually ? It's already done by the olm machine ?

**Comment by florianduros** [comment]
Abandoned: see https://github.com/vector-im/element-web/issues/25648
